samza-core/src/main/java/org/apache/samza/storage/TaskStorageCommitManager.java

```diff
@@ -0,0 +1,355 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.filefilter.WildcardFileFilter;
+import org.apache.samza.Partition;
+import org.apache.samza.SamzaException;
+import org.apache.samza.checkpoint.Checkpoint;
+import org.apache.samza.checkpoint.CheckpointId;
+import org.apache.samza.checkpoint.CheckpointManager;
+import org.apache.samza.checkpoint.CheckpointV1;
+import org.apache.samza.checkpoint.CheckpointV2;
+import org.apache.samza.checkpoint.kafka.KafkaChangelogSSPOffset;
+import org.apache.samza.config.Config;
+import org.apache.samza.container.TaskName;
+import org.apache.samza.job.model.TaskMode;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.util.FutureUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handles the commit of the state stores of the task.
+ */
+public class TaskStorageCommitManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TaskStorageCommitManager.class);
+
+  private final TaskName taskName;
+  private final CheckpointManager checkpointManager;
+  private final ContainerStorageManager containerStorageManager;
+  private final Map<String, TaskBackupManager> stateBackendToBackupManager;
+  private final Partition taskChangelogPartition;
+  private final StorageManagerUtil storageManagerUtil;
+  private final ExecutorService backupExecutor;
+  private final File durableStoreBaseDir;
+  private final Map<String, SystemStream> storeChangelogs;
+
+  // Available after init(), since stores are created by ContainerStorageManager#start()
+  private Map<String, StorageEngine> storageEngines;
+
+  public TaskStorageCommitManager(TaskName taskName, Map<String, TaskBackupManager> stateBackendToBackupManager,
+      ContainerStorageManager containerStorageManager, Map<String, SystemStream> storeChangelogs, Partition changelogPartition,
+      CheckpointManager checkpointManager, Config config, ExecutorService backupExecutor,
+      StorageManagerUtil storageManagerUtil, File durableStoreBaseDir) {
+    this.taskName = taskName;
+    this.containerStorageManager = containerStorageManager;
+    this.stateBackendToBackupManager = stateBackendToBackupManager;
+    this.taskChangelogPartition = changelogPartition;
+    this.checkpointManager = checkpointManager;
+    this.backupExecutor = backupExecutor;
+    this.durableStoreBaseDir = durableStoreBaseDir;
+    this.storeChangelogs = storeChangelogs;
+    this.storageManagerUtil = storageManagerUtil;
+  }
+
+  public void init() {
+    // Assuming that container storage manager has already started and created to stores
+    storageEngines = containerStorageManager.getAllStores(taskName);
+    if (checkpointManager != null) {
```

Comment:
It could be null if the checkpoint manager is not configured, similar null checks are in other parts of the code base such as containerStorageManager.